### PR TITLE
Fix hasChildren/hasSiblings type-hints and default value

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
follow up to #13474

## Additional info  
When a folder has only unpublished documents then it doesn't show children on tree.

Steps to reproduce:
1. Add folder under /Documents
2. Add new unpublished document under new folder
3. Reload folder => no children.

### CODE REVIEW

File: models/Document.php [hunk 1]
```diff
@@ -571,7 +571,7 @@ public function getChildren(bool $includingUnpublished = false): Listing
     /**
      * Returns true if the document has at least one child
      */
-    public function hasChildren(bool $includingUnpublished = false): bool
+    public function hasChildren(bool $includingUnpublished = null): bool
```
Comment (jdreesen) [inline models/Document.php]: How can `null` be a correct value for a boolean?

Comment (blankse) [reply models/Document.php]: https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.nullable
> It is also possible to achieve nullable arguments by making null the default value. This is not recommended as if the default value is changed in a child class a type compatibility violation will be raised as the null type will need to be added to the type declaration.

Comment (mattamon) [reply models/Document.php]: @dvesh3 shouldn't we make it true by default and not null? Like @blankse mentioned null could have side effects for child classes.

Null and true deliver the same results right?

**Review by mattamon — CHANGES_REQUESTED** [review]

File: models/Document.php (hunk 1 above)
Comment (jdreesen) [reply models/Document.php]: Yes, I know that it's technically possible. My question was about why do we *want* it to be `null` here? Why is `true` and `false` not enough, why do we need a third state? Maybe that wasn't clear enough :wink:

Comment (blankse) [reply models/Document.php]: @mattamon No, we should make a ? sign before bool. null and true is not the same.

Comment (blankse) [reply models/Document.php]: `null` use the default behaviour from `Model\Document::doHideUnpublished()`
https://github.com/pimcore/pimcore/blob/64e31dba49c1a5c8ee584f2e30efc6834ce5eb16/models/Document/Dao.php#L352-L354

Comment (mattamon) [reply models/Document.php]: @blankse good catch, thank you. For more clarity I would use only true and false like @jdreesen mentioned. I am not sure if `hasChildren` should use the default behaviour from 'Model\Document' but I guess it makes sense if you want to check if an element has only published children if you set it in `Model\Document`

Comment (mattamon) [inline models/Document.php]:
```suggestion
    public function hasChildren(?bool $includingUnpublished = null): bool
```

File: models/Document/Hardlink.php [hunk 2]
```diff
@@ -194,7 +194,7 @@ public function getChildren(bool $includingUnpublished = false): Listing
     /**
      * {@inheritdoc}
      */
-    public function hasChildren(bool $includingUnpublished = false): bool
+    public function hasChildren(bool $includingUnpublished = null): bool
```
Comment (mattamon) [inline models/Document/Hardlink.php]:
```suggestion
    public function hasChildren(?bool $includingUnpublished = null): bool
```

File: models/Document/Hardlink/Wrapper.php [hunk 3]
```diff
@@ -156,7 +156,7 @@ public function getChildren(bool $includingUnpublished = false): Listing
         return $this->children[$cacheKey];
     }
 
-    public function hasChildren(bool $includingUnpublished = false): bool
+    public function hasChildren(bool $includingUnpublished = null): bool
```
Comment (mattamon) [inline models/Document/Hardlink/Wrapper.php]:
```suggestion
    public function hasChildren(?bool $includingUnpublished = null): bool
```

**Review by mattamon — CHANGES_REQUESTED** [review]
Make the params nullable

Do we need to change all other methods too then?

File: models/Document.php (hunk 1 above)
Comment (dvesh3) [reply models/Document.php]: I would not change it to `true` as it changes the default behavior like blankse shared. Moreover, we should keep it consistent with https://github.com/pimcore/pimcore/blob/c26aaed6e70b90e735951ad6d69a04698af4aec7/models/DataObject/AbstractObject.php#L386

Comment (blankse) [reply models/Document.php]: @mattamon It is the behaviour of 10.5, see:
https://github.com/pimcore/pimcore/blob/a4117973b09b0177f7f04ddf2e91930d5a7452cb/models/Document.php#L676
Why do you want to change it?

**Review by mattamon — APPROVED** [review]
LGTM now.
If these params are now nullable maybe we should change it in AbtractObject too?

File: models/Document.php (hunk 1 above)
Comment (jdreesen) [reply models/Document.php]: @blankse got it. You are right, `null` is necessary here.

File: models/DataObject/AbstractObject.php [hunk 4]
```diff
@@ -424,7 +428,7 @@ public function getSiblings(
      */
     public function hasSiblings(
         array $objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJECT_TYPE_VARIANT, self::OBJECT_TYPE_FOLDER],
```
Comment (jdreesen) [inline models/DataObject/AbstractObject.php]:
```suggestion
        array $objectTypes = [
            self::OBJECT_TYPE_OBJECT,
            self::OBJECT_TYPE_VARIANT,
            self::OBJECT_TYPE_FOLDER
        ],
```

**Pull request merged**